File: algorithms/medium/apply_discount_every_n_orders.ex
defmodule Cashier do
  use GenServer

  @spec init_(n :: integer, discount :: integer, products :: [integer], prices :: [integer]) ::
          any
  def init_(n, discount, products, prices) do
    GenServer.start_link(__MODULE__, [], name: __MODULE__)

    prices =
      Enum.zip(products, prices)
      |> Enum.reduce(%{}, fn {i, p}, acc ->
        Map.put(acc, i, p)
      end)

    GenServer.call(__MODULE__, {:init, {n, 0, discount, prices}})
  end

  @spec get_bill(product :: [integer], amount :: [integer]) :: float
  def get_bill(product, amount) do
    GenServer.call(__MODULE__, {:get_bill, product, amount})
  end

  def init(state) do
    {:ok, state}
  end

  def handle_call({:init, state}, _from, _state) do
    {:reply, true, state}
  end

  def handle_call({:get_bill, products, amounts}, _from, {n, ct, discount, prices}) do
    bill =
      Enum.zip(products, amounts)
      |> Enum.reduce(0, fn {i, amt}, acc -> acc + prices[i] * amt end)

    ct = ct + 1

    if rem(ct, n) == 0 do
      {:reply, bill * (1 - discount / 100), {n, ct, discount, prices}}
    else
      {:reply, bill, {n, ct, discount, prices}}
    end
  end
end

File: algorithms/medium/apply_discount_every_n_orders.rs
struct Cashier {
    n: i32,
    ct: i32,
    discount: f64,
    prices: Vec<f64>,
}

impl Cashier {

    fn new(n: i32, discount: i32, products: Vec<i32>, prices: Vec<i32>) -> Self {
        let mut prices_ = vec![0.0; 201];
        products.into_iter().zip(prices).for_each(|(i, p)| prices_[i as usize] = p as f64);
        Self {
            n,
            ct: 0,
            discount: discount as f64,
            prices: prices_,
        }
    }
    
    fn get_bill(&mut self, product: Vec<i32>, amount: Vec<i32>) -> f64 {
        let mut bill = product.into_iter()
            .zip(amount)
            .fold(0.0, |acc, (i, ct)| acc + self.prices[i as usize] * (ct as f64));
        
        self.ct += 1;
        
        if self.ct % self.n == 0 {
            bill * (1.0 - self.discount / 100.0)
        } else {
            bill
        }
    }
}
